### PR TITLE
Extract package metadata info + refactor

### PR DESCRIPTION
**Pull request by piotmag769**
commit-id:b9eb7559

---

**Stack**:
- #1464
- #1463
- #1462
- #1461
- #1460 ⬅

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

### CODE REVIEW

File: extensions/scarb-doc/src/lib.rs [hunk 1]
```diff
@@ -1,16 +1,56 @@
...
+) -> Vec<PackageInformation> {
+    let mut packages_information = vec![];
+    for package_metadata in metadata_for_packages {
+        let authors = package_metadata.manifest_metadata.authors.clone();
+
+        let project_config = get_project_config(metadata, package_metadata);
+        let crate_ = generate_language_elements_tree_for_package(
+            package_metadata.name.clone(),
+            project_config,
+        );
```
Comment (piotmag769) [inline extensions/scarb-doc/src/lib.rs:40]: Moved from `main.rs`

**Review by mkaput — APPROVED** [review]

File: extensions/scarb-doc/src/versioned_json_output.rs [hunk 2]
```diff
@@ -1,22 +1,38 @@
...
+use std::fs;
 
 const FORMAT_VERSION: u8 = 1;
+const JSON_OUTPUT_FILENAME: &str = "output.json";
 
 #[derive(Serialize)]
 pub struct VersionedJsonOutput {
     format_version: u8,
-    pub package_information_map: BTreeMap<PackageName, Crate>,
+    pub package_information_map: Vec<PackageInformation>,
```
Comment (Arcticae) [inline extensions/scarb-doc/src/versioned_json_output.rs:13]: Outdated name

File: extensions/scarb-doc/src/versioned_json_output.rs [hunk 3]
```diff
@@ -1,22 +1,38 @@
...
 }
 
 impl VersionedJsonOutput {
-    pub fn new(package_information_map: BTreeMap<PackageName, Crate>) -> Self {
+    pub fn new(package_information_map: Vec<PackageInformation>) -> Self {
```
Comment (Arcticae) [inline extensions/scarb-doc/src/versioned_json_output.rs:17]: Outdated name

File: extensions/scarb-doc/tests/data/json_output_test_data.json [hunk 4]
```diff
@@ -1,281 +1,287 @@
...
-            "free_functions": [
-              {
-                "item_data": {
-                  "name": "it_works",
-                  "doc": "Really\nworks.\n",
-                  "signature": "fn it_works()",
-                  "full_path": "hello_world::tests::it_works"
+  "package_information_map": [
+    {
+      "crate_": {
```
Comment (Arcticae) [inline extensions/scarb-doc/tests/data/json_output_test_data.json:5]: Why is this name named like this?

**Review by Arcticae — APPROVED** [review]
Few nitpicks

File: extensions/scarb-doc/tests/data/json_output_test_data.json (hunk 4 above)
Comment (piotmag769) [reply extensions/scarb-doc/tests/data/json_output_test_data.json:5]: `crate` is a keyword in Rust

**Pull request merged**